### PR TITLE
make x86_64-unknown-linux-gnu ubuntu again

### LINKED ISSUE #680: glibc compatibility / make ` x86_64-unknown-linux-gnu` ubuntu again and provide image for glibc compatibility
I think it's unexpected all linux images are ubuntu, except `x86_64-unknown-linux-gnu`, I think this can be remedied by building two images: `x86_64-unknown-linux-gnu` and ` x86_64-unknown-linux-gnu.2.17`

Thankfully, we have not published the centos change yet, so this can be remedied

Related to the glibc issue, I think we can remedy this by suggesting users use `zig cc` ala [cargo-zigbuild](https://github.com/messense/cargo-zigbuild). We can even use a similar approach setting `*_CC` to `zig cc --target x86_64-linux-gnu.2.17` and make it easier for users of the `x86_64-unknown-linux-gnu` image to choose a glibc

ping #591

### PR DESCRIPTION
**Pull request by Emilgardis**
This pr effectively changes the default image for `x86_64-unknown-linux-gnu` to be ubuntu based.

We target 16.04 because that is what was used in cross 0.2.1.

While some users may not care about this, if GLIBC 2.23 is too new for someone, you're able to replace the image in `Cross.toml` with a CentOS based one (targets 2.17):

```
[target.x86_64-unknown-linux-gnu]
# The CentOS image gets published under the same package as the Ubuntu one, just with an added `-centos`
image = "ghcr.io/cross-rs/cross/x86_64-unknown-linux-gnu:main-centos"
```

Resolves #680

### CODE REVIEW

File: build-docker-image.sh [hunk 1]
```diff
@@ -20,15 +20,17 @@ pushd docker
 run() {
   local push="${1}"
   local build_args=()
-
+  echo "${1} - ${2}"
```
Comment (Emilgardis) [inline build-docker-image.sh:23]: oops

**Comment by Emilgardis** [comment]
bors try --target powerpc-unknown-linux-gnu --target x86_64-unknown-linux-gnu.2.17

**Comment by Emilgardis** [comment]
that's annoying...

**Comment by Emilgardis** [comment]
bors cancel

**Comment by Emilgardis** [comment]
bors try --target powerpc-unknown-linux-gnu --target x86_64-unknown-linux-gnu.2.17

**Comment by Emilgardis** [comment]
bors cancel

**Comment by Emilgardis** [comment]
`yq --output-format json '[.[] | select(.target + (if .sub then "." + .sub else "" end) == "powerpc-unknown-linux-gnu" or .target + (if .sub then "." + .sub else "" end) == "x86_64-unknown-linux-gnu.2.17")]'` works on my machine, interesting...

**Comment by Emilgardis** [comment]
aha, yq is https://github.com/mikefarah/yq, not the python one

**Comment by Emilgardis** [comment]
bors cancel

bash is not my friend

**Comment by Emilgardis** [comment]
bors try --target powerpc-unknown-linux-gnu --target x86_64-unknown-linux-gnu.2.17

**Comment by Emilgardis** [comment]
bors cancel

**Comment by Emilgardis** [comment]
bors try --target powerpc-unknown-linux-gnu --target x86_64-unknown-linux-gnu.2.17

**Comment by Emilgardis** [comment]
bors ping

bors try --target powerpc-unknown-linux-gnu --target x86_64-unknown-linux-gnu.2.17

**Comment by Emilgardis** [comment]
uninstalled bors and reinstalled.

bors try --target powerpc-unknown-linux-gnu --target x86_64-unknown-linux-gnu.2.17

**Comment by Alexhuszagh** [comment]
Relating to #680, the only issue with `zig cc` is the lack of C++ support right now: I think just building glibc from source and then GCC might be the best option.

**Comment by Alexhuszagh** [comment]
Oh also, this PR would also need to include the Ubuntu version of `x86_64-unknown-linux-gnu` in addition.

**Comment by Alexhuszagh** [comment]
The CentOS7 image can't support bindgen, since the clang is too old from the official repositories, which I think is an issue. I think we either need to build our own toolchain or provide 2 toolchains, 1 CentOS-based and 1 Ubuntu-based, with the Ubuntu-based using a more recent toolchain.

File: build-docker-image.sh [hunk 2]
```diff
@@ -20,15 +20,17 @@ pushd docker
...
     build_args+=(--push)
   else
     build_args+=(--load)
   fi
 
   local dockerfile="Dockerfile.${2}"
-  local image_name="ghcr.io/cross-rs/${2}"
+  local name=${2%%.*}
+  local suffix=${2#*.}
+  local image_name="ghcr.io/cross-rs/${name}"
```
Comment (Emilgardis) [inline build-docker-image.sh:33]: this doesn't work

```bash
# Doesn't work
two="x86_64-unknown-linux-gnu"
echo ${two%%.*} # x86_64-unknown-linux-gnu
echo ${two#*.} # x86_64-unknown-linux-gnu

# Works
two="x86_64-unknown-linux-gnu.2.17"
echo ${two%%.*} # x86_64-unknown-linux-gnu
echo ${two#*.} # 2.17
```

not sure how to fix

Comment (Emilgardis) [reply build-docker-image.sh:33]: #786 will help! :D

**Comment by Emilgardis** [comment]
this should be good to go 🎉

File: README.md [hunk 3]
```diff
@@ -318,7 +318,7 @@ terminate.
 | `x86_64-pc-windows-gnu`              | N/A    | 7.3     | ✓   | N/A   |   ✓    |
 | `x86_64-unknown-freebsd`             | 1.5    | 6.4.0   | ✓   | N/A   |       |
 | `x86_64-unknown-dragonfly` [2] [3]   | 6.0.1  | 5.3.0   | ✓   | N/A   |       |
-| `x86_64-unknown-linux-gnu`           | 2.17   | 4.8.5   | ✓   | 4.2.1 |   ✓    |
+| `x86_64-unknown-linux-gnu`           | 2.23   | 5.4.0   | ✓   | 5.1.0 |   ✓    |
```
Comment (Alexhuszagh) [inline README.md:321]: We should add documentation for the `x86_64-unknown-linux-gnu:centos` image as well.

File: docker/Dockerfile.x86_64-unknown-linux-gnu [hunk 4]
```diff
@@ -22,8 +16,9 @@ RUN /qemu.sh x86_64 softmmu
...
 RUN /dropbear.sh
 
-COPY --from=0 /qemu /qemu
+COPY linux-image.sh /
+RUN /linux-image.sh x86_64
 
 COPY linux-runner /
 
-ENV CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_RUNNER="/linux-runner x86_64"
+ENV CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_RUNNER="/linux-runner x86_64"
```
Comment (Alexhuszagh) [inline docker/Dockerfile.x86_64-unknown-linux-gnu:24]: I don't think this change was intentional, also, someone's text editor will automatically add this newline.

**Comment by Alexhuszagh** [comment]
Looks good, outside the suggestion we should document the CentOS image, since glibc compatibility seems to be a pretty popular request from some users.

**Comment by Emilgardis** [comment]
I'm not sure how to document it in the table. Maybe a new column or just a special entry an asterisk

**Comment by Alexhuszagh** [comment]
> I'm not sure how to document it in the table. Maybe a new column or just a special entry an asterisk

Something like this:

```markdown
| Target                               |  libc  |   GCC   | C++ | QEMU  | `test` |
|--------------------------------------|-------:|--------:|:---:|------:|:------:|
| `x86_64-unknown-linux-gnu:centos` [5]  | 2.17   | 4.8.5   | ✓   | 4.2.1 |   ✓    |

[5] Must change the image in `Cross.toml` for `[target.x86_64-unknown-linux-gnu]`.
```

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -102,15 +103,15 @@ jobs:
 
           selects=()
           for target in "${targets[@]}"; do
-            selects+=(".target == \"${target}\"")
+            selects+=(".target + (if .sub then \".\" + .sub else \"\" end) == \"${target}\"")
```
[the remark this answers is not in the document]
Comment (Emilgardis) [reply .github/workflows/ci.yml:105]: where would I get this value from?

**Comment by Emilgardis** [comment]
bors try --target x86_64-unknown-linux-gnu --target x86_64-unknown-linux-gnu.centos

this try will be unusually long as we don't have a cache for these

File: .github/workflows/ci.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (Alexhuszagh) [reply .github/workflows/ci.yml:105]: Actually I think it's fine, it's probably better than the other jq syntax.

**Review by Alexhuszagh — APPROVED** [review]

**Comment by Alexhuszagh** [comment]
> are there anymore issues that this fixes?

Just #680 I believe.

**Comment by Emilgardis** [comment]
I see one more thing to fix:

image

**Comment by Emilgardis** [comment]
bors try --target x86_64-unknown-linux-gnu --target x86_64-unknown-linux-gnu.centos

**Comment by Emilgardis** [comment]
canceled it early, but it should work

https://github.com/cross-rs/cross/runs/6889886219?check_suite_focus=true#step:2:125

**Comment by Emilgardis** [comment]
bors try --target x86_64-unknown-linux-gnu --target x86_64-unknown-linux-gnu.centos

**Pull request merged**